activerecord/lib/active_record/relation/spawn_methods.rb

```diff
@@ -37,7 +38,7 @@ def merge(other)
     end
 
     def merge!(other) # :nodoc:
-      if !other.is_a?(Relation) && other.respond_to?(:to_proc)
+      if other.is_a?(Proc)
```

Comment:
I like @pixeltrix's implementation. :+1: for that even if we don't need to support duck typing. It is cleaner than the actual implementation.